qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,828 @@
+                    ),
+                    SeriesDef(
+                        name="y",
+                        param_names=["a", "freq", "phase", "b"],
+                        fit_func_index=1,
+                        filter_kwargs={"pulse": "y"},
+                        data_option_keys=["outcome"]
+                    )
+                ]
+
+                __fit_funcs__ = [library.cos, library.sin]
+
+                __param_names__ = ["a", "freq", "phase", "b"]
+
+    Notes:
+        This CurveAnalysis class provides several private methods that subclasses can override.
+
+        - Customize figure generation:
+            Override :meth:`~self._create_figures`. For example, here you can create
+            arbitrary number of new figures or upgrade the default figure appearance.
+
+        - Customize pre-data processing:
+            Override :meth:`~self._data_pre_processing`. For example, here you can
+            take a mean over y values for the same x value, or apply smoothing to y values.
+
+        - Customize post-analysis data processing:
+            Override :meth:`~self._post_processing`. For example, here you can
+            calculate new entity from fit values. Such as EPC of RB experiment.
+
+        - Customize fitting options:
+            Override :meth:`~self._setup_fitting`. For example, here you can
+            calculate initial guess from experiment data and setup fitter options.
+
+        - Customize data processor calibration:
+            Override :meth:`~Self._calibrate_data_processor`. This is special subroutine
+            that is only called when a DataProcessor instance is used as the data processor.
+            You can take arbitrary data from experiment result and setup your processor.
+
+        Note that other private methods are not expected to be overridden.
+        If you forcibly override these methods, the behavior of analysis logic is not well tested
+        and we cannot guarantee it works as expected (you may suffer from bugs).
+        Instead, you can open an issue in qiskit-experiment github to upgrade this class
+        with proper unittest framework.
+
+        https://github.com/Qiskit/qiskit-experiments/issues
+    """
+
+    #: str: Metadata key representing a scanned value.
+    __x_key__ = "xval"
+
+    #: List[SeriesDef]: List of mapping representing a data series
+    __series__ = None
+
+    #: List[Callable]: A callback function to define the expected curve
+    __fit_funcs__ = None
+
+    #: List[str]: Parameter name list
+    __param_names__ = list()
+
+    # Callable: Default curve fitter. This can be overwritten.
+    __base_fitter__ = scipy_curve_fit_wrapper
+
+    # Union[Callable, DataProcessor]: Data processor to format experiment data.
+    __default_data_processor__ = level2_probability
+
+    # pylint: disable = unused-argument, missing-return-type-doc
+    def _create_figures(self, curve_data: List[CurveEntry], fit_data: AnalysisResult):
```

Comment:
Is `curve_data` general enough for RB plots which have both the raw x, y data, and the averaged x,y, sigma data (and raw series and averaged series for irb)?

Reply:
Yes, I think so. The raw data contains un-averaged samples for each series (i.e. standard and not interleaved) with circuit metadata. So we need to splint the sample data into curve of each seed (because seed info is not preserved).